go/vt/vtgate/planbuilder/physical/operator_funcs.go

```diff
@@ -0,0 +1,363 @@
+			if i > -1 {
+				continue
+			}
+			op.Columns = append(op.Columns, col)
+			noQualifierNames = append(noQualifierNames, sqlparser.NewColName(col.Name.String()))
+		}
+		if len(noQualifierNames) > 0 {
+			_, _, err := PushOutputColumns(ctx, op.Source, noQualifierNames...)
+			if err != nil {
+				return nil, nil, err
+			}
+		}
+		return op, offsets, nil
+
+	default:
+		return nil, nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "we cannot push output columns into %T", op)
+	}
+}
+
+func addToIntSlice(columnOffset []int, valToAdd int) ([]int, int) {
+	for idx, val := range columnOffset {
+		if val == valToAdd {
+			return columnOffset, idx
+		}
+	}
+	columnOffset = append(columnOffset, valToAdd)
+	return columnOffset, len(columnOffset) - 1
+}
+
+func RemovePredicate(ctx *plancontext.PlanningContext, expr sqlparser.Expr, op abstract.PhysicalOperator) (abstract.PhysicalOperator, error) {
+	switch op := op.(type) {
+	case *Route:
+		newSrc, err := RemovePredicate(ctx, expr, op.Source)
+		if err != nil {
+			return nil, err
+		}
+		op.Source = newSrc
+		return op, err
+	case *ApplyJoin:
+		isRemoved := false
+		deps := ctx.SemTable.RecursiveDeps(expr)
+		if deps.IsSolvedBy(op.LHS.TableID()) {
+			newSrc, err := RemovePredicate(ctx, expr, op.LHS)
+			if err != nil {
+				return nil, err
+			}
+			op.LHS = newSrc
+			isRemoved = true
+		}
+
+		if deps.IsSolvedBy(op.RHS.TableID()) {
+			newSrc, err := RemovePredicate(ctx, expr, op.RHS)
+			if err != nil {
+				return nil, err
+			}
+			op.RHS = newSrc
+			isRemoved = true
+		}
+
+		var keep []sqlparser.Expr
+		for _, e := range sqlparser.SplitAndExpression(nil, op.Predicate) {
+			if !sqlparser.EqualsExpr(expr, e) {
+				keep = append(keep, e)
+				isRemoved = true
+			}
+		}
+		op.Predicate = sqlparser.AndExpressions(keep...)
+
+		if !isRemoved {
+			return nil, vterrors.Errorf(vtrpcpb.Code_UNIMPLEMENTED, "remove '%s' predicate not supported on cross-shard join query", sqlparser.String(expr))
+		}
+		return op, nil
+	case *Filter:
+		idx := -1
+		for i, predicate := range op.Predicates {
+			if sqlparser.EqualsExpr(predicate, expr) {
+				idx = i
+			}
+		}
+		if idx == -1 {
+			// the predicate is not here. let's remove it from our source
+			newSrc, err := RemovePredicate(ctx, expr, op.Source)
+			if err != nil {
+				return nil, err
+			}
+			op.Source = newSrc
+			return op, nil
+		}
+		if len(op.Predicates) == 1 {
+			// no predicates left on this operator, so we just remove it
+			return op.Source, nil
+		}
+
+		// remove the predicate from this filter
+		op.Predicates = append(op.Predicates[:idx], op.Predicates[idx+1:]...)
+		return op, nil
+
+	default:
+		return nil, vterrors.Errorf(vtrpcpb.Code_INTERNAL, "this should not happen - tried to remove predicate from table op")
+	}
+}
+
+func breakExpressionInLHSandRHS(
+	ctx *plancontext.PlanningContext,
+	expr sqlparser.Expr,
+	lhs semantics.TableSet,
+) (bvNames []string, columns []*sqlparser.ColName, rewrittenExpr sqlparser.Expr, err error) {
```

Comment:
This function exists in horizon_planning too. We should export this one and use it there too instead of having 2 copies.